File: mars_rover_tdd_java/src/main/java/com/example/mars_rover_tdd_java/MarsRover.java
package com.example.mars_rover_tdd_java;

public class MarsRover {

    public int x;
    public int y;
    public String direction;

    public MarsRover(int x, int y, String direction) {
        this.x = x;
        this.y = y;
        this.direction = direction;
    }

    public int getX() {
        return this.x;
    }

    public int getY() {
        return this.y;
    }

    public String getDirection() {
        return this.direction;
    }

    public boolean validatePosition(int x, int y, String direction) {
        return ((this.x == x) && (this.y == y)  && (this.direction.equals(direction)));
    }

    public void movementOrders(String movementOrders) {
        String[] movements = movementOrders.split("");
        for (String movement : movements) {
            move(movement);
        }
    }

    public void move(String movement) {
        switch (movement.toLowerCase()) {
            case "f" -> moveForward(this.direction);
            case "b" -> moveBackward(this.direction);
            case "l" -> turnLeft(this.direction);
            case "r" -> turnRight(this.direction);
            default -> throw new IllegalStateException("Unexpected value: " + movement);
        }
    }

    public void turnLeft(String direction) {
        switch (direction) {
            case "N" -> this.direction = "W";
            case "W" -> this.direction = "S";
            case "S" -> this.direction = "E";
            case "E" -> this.direction = "N";
            default -> throw new IllegalStateException("Unexpected value: " + direction);
        }
    }

    public void turnRight(String direction) {
        switch (direction) {
            case "N" -> this.direction = "E";
            case "E" -> this.direction = "S";
            case "S" -> this.direction = "W";
            case "W" -> this.direction = "N";
            default -> throw new IllegalStateException("Unexpected value: " + direction);
        }
    }

    public void moveForward(String direction) {
        switch (direction) {
            case "N" -> this.y += 1;
            case "S" -> this.y -= 1;
            case "E" -> this.x += 1;
            case "W" -> this.x -= 1;
            default -> throw new IllegalStateException("Unexpected value: " + direction);
        }
    }

    public void moveBackward(String direction) {
        switch (direction) {
            case "N" -> this.y -= 1;
            case "S" -> this.y += 1;
            case "E" -> this.x -= 1;
            case "W" -> this.x += 1;
            default -> throw new IllegalStateException("Unexpected value: " + direction);
        }
    }
}

File: mars_rovers_tdd_csharp/mars_rovers_tdd_csharp/models/MarsRover.cs
namespace mars_rovers_tdd_csharp.models;

public class MarsRover
{
    private int _coordinateX;
    private int _coordinateY;
    private string _direction;
    
    public MarsRover(int coordinateX, int coordinateY, string direction)
    {
        this._coordinateX = coordinateX;
        this._coordinateY = coordinateY;
        this._direction = direction;
    }
    
    public int GetX()
    {
        return _coordinateX;
    }
    
    public int GetY()
    {
        return _coordinateY;
    }
    
    public string GetDirection()
    {
        return _direction;
    }
    
    public bool ValidatePosition(int x, int y, string direction)
    {
        return this._coordinateX == x && this._coordinateY == y && this._direction == direction;
    }

    public void MovementOrders(string movementOrders)
    {
        var movements = movementOrders.Select(c => c.ToString()).ToArray();
        foreach (var movement in movements)
        {
            Move(movement);
        }
    }
    
    public void Move(string movement)
    {
        switch (movement.ToLower())
        {
            case "f":
                MoveForward(_direction);
                break;
            case "b":
                MoveBackward(_direction);
                break;
            case "l":
                TurnLeft(_direction);
                break;
            case "r":
                TurnRight(_direction);
                break;
            default:
                throw new ArgumentException($"Invalid movement: {movement}");
        }
    }
    
    public void TurnLeft(string direction)
    {
        switch (direction)
        {
            case "N":
                this._direction = "W";
                break;
            case "W":
                this._direction = "S";
                break;
            case "S":
                this._direction = "E";
                break;
            case "E":
                this._direction = "N";
                break;
            default:
                throw new ArgumentException($"Invalid direction: {direction}");
        }
    }
    
    public void TurnRight(string direction)
    {
        switch (direction)
        {
            case "N":
                this._direction = "E";
                break;
            case "E":
                this._direction = "S";
                break;
            case "S":
                this._direction = "W";
                break;
            case "W":
                this._direction = "N";
                break;
            default:
                throw new ArgumentException($"Invalid direction: {direction}");
        }
    }
    
    public void MoveForward(string direction)
    {
        switch (direction)
        {
            case "N":
                _coordinateY++;
                break;
            case "E":
                _coordinateX++;
                break;
            case "S":
                _coordinateY--;
                break;
            case "W":
                _coordinateX--;
                break;
            default:
                throw new ArgumentException($"Invalid direction: {direction}");
        }
    }
    
    public void MoveBackward(string direction)
    {
        switch (direction)
        {
            case "N":
                _coordinateY--;
                break;
            case "E":
                _coordinateX--;
                break;
            case "S":
                _coordinateY++;
                break;
            case "W":
                _coordinateX++;
                break;
            default:
                throw new ArgumentException($"Invalid direction: {direction}");
        }
    }
}